File: 2327-number-of-people-aware-of-a-secret/2327-number-of-people-aware-of-a-secret.py
class Solution:
    def peopleAwareOfSecret(self, n: int, delay: int, forget: int) -> int:
        MOD = 10**9 + 7
        dp = [0] * (n + 1)
        pref = [0] * (n + 1)
        dp[1] = 1
        pref[1] = 1

        for i in range(2, n + 1):
            r = i - delay      # prefix 上界
            l = i - forget     # prefix 下界 (要用 pref[l])
            if r >= 1:
                left = pref[r]
                right = pref[l] if l >= 0 else 0
                dp[i] = (left - right) % MOD
            else:
                dp[i] = 0
            pref[i] = (pref[i-1] + dp[i]) % MOD

        # 最終答案：在第 n 天仍記得的人
        ans = (pref[n] - (pref[n - forget] if n - forget >= 0 else 0)) % MOD
        return ans

File: 2327-number-of-people-aware-of-a-secret/2327-number-of-people-aware-of-a-secret.cpp
class Solution {
public:
    int peopleAwareOfSecret(int n, int delay, int forget) {
        const int MOD = 1e9 + 7;
        vector<long long> dp(n + 1, 0);    // dp[i] = 第 i 天新知道秘密的人數
        vector<long long> pref(n + 1, 0);  // 前綴和

        dp[1] = 1;
        pref[1] = 1;

        for (int i = 2; i <= n; i++) {
            int r = i - delay;   // 可開始傳播的最晚天數
            int l = i - forget;  // 已經忘記的最早天數

            if (r >= 1) {
                long long left = pref[r];
                long long right = (l >= 0 ? pref[l] : 0);
                dp[i] = (left - right + MOD) % MOD;
            } else {
                dp[i] = 0;
            }
            pref[i] = (pref[i - 1] + dp[i]) % MOD;
        }

        long long ans = (pref[n] - (n - forget >= 0 ? pref[n - forget] : 0) + MOD) % MOD;
        return (int)ans;
    }
};